lib/active_merchant/billing/gateways/omise.rb

```diff
@@ -0,0 +1,371 @@
+      # * <tt>:secret_key</tt> -- Omise's secret key (REQUIRED).
+
+      def initialize(options={})
+        requires!(options, :public_key, :secret_key)
+        @public_key = options[:public_key]
+        @secret_key = options[:secret_key]
+        super
+      end
+
+      # Perform a purchase.
+      #
+      # ==== Parameters
+      #
+      # * <tt>money</tt>   -- The purchasing amount as the value is in Satang.
+      # * <tt>payment</tt> -- The CreditCard object or card token that is used for the transaction.
+      # * <tt>options</tt> -- An optional parameters.
+      #
+      # ==== Options
+      # * <tt>token_id</tt> -- token id, use Omise.js library to retrieve a token id
+      # if this is passed as an option, it will ignore tokenizing via Omisevaultgateway object
+      #
+      # === Example
+      #  To create a charge on a card
+      #
+      #   purchase(money, creditcard_object)
+      #
+      #  To create a charge on a token
+      #
+      #   purchase(money, nil, { :token_id => token_id, ... })
+      #
+      #  To create a charge on a customer
+      #
+      #   purchase(money, nil, { :customer_id => customer_id })
+
+      def purchase(money, payment, options={})
+        create_post_for_purchase(money, payment, options)
+      end
+
+      # Authorize a charge.
+      #
+      # ==== Parameters
+      #
+      # * <tt>money</tt>   -- An amount of money to charge in Satang.
+      # * <tt>payment</tt> -- The CreditCard
+      # * <tt>options</tt> -- A standard hash options
+
+      def authorize(money, payment, options={})
+        options[:capture] = 'false'
+        create_post_for_purchase(money, payment, options)
+      end
+
+      # Capture or a pre-authorized charge.
+      #
+      # ==== Parameters
+      #
+      # * <tt>money</tt>     -- An amount of money to charge in Satang.
+      # * <tt>charge_id</tt> -- The Charge object identifier.
+      # * <tt>options</tt>   -- A standard hash options.
+
+      def capture(money, charge_id, options={})
+        post = {}
+        add_amount(post, money, options)
+        commit(:post, "charges/#{CGI.escape(charge_id)}/capture", post, options)
+      end
+
+      # Void a charge.
+      #
+      # ==== Parameters
+      #
+      # * <tt>charge_id</tt> -- The Charge object identifier.
+      # * <tt>options</tt>   -- A standard hash options.
+
+      def void(charge_id, options={})
+        MultiResponse.run(:first) do |r|
+          r.process { commit(:get, "charges/#{CGI.escape(charge_id)}") }
+          r.process { refund(r.params['amount'], charge_id, options) }
+        end.responses.last
+      end
+
+      # Refund a charge.
+      #
+      # ==== Parameters
+      #
+      # * <tt>money</tt>     -- An amount of money to charge in Satang.
+      # * <tt>charge_id</tt> -- The Charge object identifier.
+      # * <tt>options</tt>   -- A standard hash options.
+
+      def refund(money, charge_id, options={})
+        options[:amount] = money if money
```

Comment:
@girasquid well, with many adapters you can pass in `nil` as the amount and it'll just refund the whole transaction. That said, unless having a blank `amount` blows up transactions, you still shouldn't need the check.